crates/next-api/src/module_graph.rs

```diff
@@ -0,0 +1,660 @@
+            //     .iter()
+            //     .map(|issue| issue.to_resolved())
+            //     .try_join()
+            // .await?;
+
+            let edges = refs
+                .iter()
+                .flat_map(|(ty, modules)| {
+                    if matches!(ty, ChunkingType::Traced) {
+                        None
+                    } else {
+                        Some(modules.iter())
+                    }
+                })
+                .flatten()
+                .map(|m| async move {
+                    Ok(SingleModuleGraphBuilderEdge {
+                        to: SingleModuleGraphBuilderNode::new(*m).await?,
+                    })
+                })
+                .try_join()
+                .await?;
+            // if !refs_issues.is_empty() {
+            //     x.push(SingleModuleGraphBuilderEdge {
+            //         to: SingleModuleGraphBuilderNode::Issues(refs_issues),
+            //     });
+            // }
+            Ok(edges)
+        }
+    }
+
+    fn span(&mut self, node: &SingleModuleGraphBuilderNode) -> tracing::Span {
+        match node {
+            SingleModuleGraphBuilderNode::Module { ident, .. } => {
+                tracing::info_span!("module", name = display(ident))
+            }
+            SingleModuleGraphBuilderNode::Issues(_) => {
+                tracing::info_span!("issues")
+            }
+        }
+    }
+}
+
+impl SingleModuleGraph {
+    /// Walks the graph starting from the given entries and collects all reachable nodes, skipping
+    /// nodes listed in `visited_modules`
+    /// If passed, `root` is connected to the entries and include in `self.entries`.
+    async fn new_inner(
+        root: Option<ResolvedVc<Box<dyn Module>>>,
+        entries: &Vec<ResolvedVc<Box<dyn Module>>>,
+        visited_modules: &HashSet<ResolvedVc<Box<dyn Module>>>,
+    ) -> Result<Vc<Self>> {
+        let mut graph = DiGraph::new();
+
+        let root_edges = entries
+            .iter()
+            .map(|e| async move {
+                Ok(SingleModuleGraphBuilderEdge {
+                    to: SingleModuleGraphBuilderNode::new(*e).await?,
+                })
+            })
+            .try_join()
+            .await?;
+        let children_modules_iter = AdjacencyMap::new()
+            .skip_duplicates_with_visited_nodes(VisitedNodes(
+                visited_modules
+                    .iter()
+                    .map(|&module| SingleModuleGraphBuilderNode::new(module))
+                    .try_join()
+                    .await?
+                    .into_iter()
+                    .collect(),
+            ))
+            .visit(root_edges, SingleModuleGraphBuilder {})
+            .await
+            .completed()?
+            .into_inner();
+
+        let mut modules: HashMap<ResolvedVc<Box<dyn Module>>, NodeIndex<u32>> = HashMap::new();
+        {
+            let _span = tracing::info_span!("build petgraph").entered();
```

Comment:
A info_span is potentially shown to the end user, so we probably don't want to include details like "petgraph".

```suggestion
            let _span = tracing::info_span!("build module graph").entered();
```